jib-core/src/main/java/com/google/cloud/tools/jib/image/ImageLayers.java

```diff
@@ -19,34 +19,31 @@
 import com.google.common.collect.ImmutableList;
 import com.google.common.collect.ImmutableSet;
 import java.util.Iterator;
+import java.util.LinkedHashSet;
 import javax.annotation.Nullable;
 
-/** Holds the layers for an image. Makes sure that each layer is only added once. */
+/** Holds the layers for an image. Makes sure that there are no duplicate layers. */
 public class ImageLayers<T extends Layer> implements Iterable<T> {
 
   public static class Builder<T extends Layer> {
 
-    private final ImmutableList.Builder<T> layersBuilder = ImmutableList.builder();
+    private final LinkedHashSet<T> layers = new LinkedHashSet<>();
```

Comment:
I think explicitly marking this a `LinkedHashSet` is a good idea, but it warrants a bit of explanation that the set aspect is for the fast comparison.  TBH given the need to remove-and-then-add, it would be better to just make this a List. Docker is limited to 40 layers (I believe), so the O(n) vs O(ln n) doesn't matter.

Reply:
I think the speed here might actually matter a bit, since this class is used to represent the layers in the cache as well (as this is mostly fixing #721). Having the remove-then-add would be `O(n)` performed `O(n)` times (for `n` layers). Since the cache could potentially have many layers after extensive use (not limited to the Docker image layer limit), the `O(n2)` vs `O(n ln n)` could potentially be a noticeable difference when manipulating the cache metadata.